src/lib/Tests/Form/DataTransformer/VersionInfoTransformerTest.php

```diff
@@ -0,0 +1,184 @@
+                    'version_no' => self::EXAMPLE_VERSION_NO,
+                ],
+            ],
+        ];
+    }
+
+    /**
+     * @dataProvider dataProviderForTransformWithInvalidInput
+     */
+    public function testTransformWithInvalidInput($value): void
+    {
+        $this->expectException(TransformationFailedException::class);
+        $this->expectExceptionMessage('Value cannot be transformed because the passed value is not a VersionInfo object');
+
+        $this->transformer->transform($value);
+    }
+
+    public function dataProviderForTransformWithInvalidInput(): array
+    {
+        $object = new class() {
+        };
+
+        return [
+            'string' => ['string'],
+            'integer' => [123456],
+            'bool' => [true],
+            'float' => [12.34],
+            'array' => [[]],
+            'object' => [$object],
+        ];
+    }
+
+    /**
+     * @dataProvider dataProviderForReverseTransformWithValidInput
+     */
+    public function testReverseTransformWithValidInput(?array $value, ?VersionInfo $expected): void
+    {
+        if ($expected !== null) {
+            $this->contentService
+                ->expects($this->once())
+                ->method('loadVersionInfo')
+                ->with(
+                    $this->equalTo($value['content_info']),
+                    $this->logicalAnd(
+                        $this->equalTo($value['version_no']),
+                        // Make sure value is casted to int
+                        $this->isType('int')
+                    )
+                )
+                ->willReturn($expected);
+        }
+
+        $this->assertEquals(
+            $expected,
+            $this->transformer->reverseTransform($value)
+        );
+    }
+
+    public function dataProviderForReverseTransformWithValidInput(): array
+    {
+        $contentInfo = new ContentInfo([
+            'id' => self::EXAMPLE_CONTENT_ID,
+        ]);
+
+        $versionInfo = $this->createVersionInfoMock($contentInfo, self::EXAMPLE_VERSION_NO);
+
+        return [
+            'null' => [null, null],
+            'empty' => [
+                [
+                    'content_info' => null,
+                    'version_no' => null,
+                ],
+                null,
+            ],
+            'non_empty' => [
+                [
+                    'content_info' => $contentInfo,
+                    'version_no' => self::EXAMPLE_VERSION_NO,
+                ],
+                $versionInfo,
+            ],
+            'non_empty_with_version_cast' => [
+                [
+                    'content_info' => $contentInfo,
+                    'version_no' => (string)self::EXAMPLE_VERSION_NO,
+                ],
+                $versionInfo,
+            ],
+        ];
+    }
+
+    /**
+     * @dataProvider dataProviderForReverseTransformWithInvalidInput
+     */
+    public function testReverseTransformWithInvalidInput($value): void
```

Comment:
missing strict type

Reply:
It's mixed on propose. See test cases in the related data provider.

Reply:
> On purpose there is no type spec for `$value` parameter. See test cases in the related data provider.

ref previous comment.